topdown/time.go

```diff
@@ -103,16 +104,64 @@ func builtinClock(a ast.Value) (ast.Value, error) {
 }
 
 func builtinWeekday(a ast.Value) (ast.Value, error) {
-	t, err := utcTime(a)
+	t, err := tzTime(a)
 	if err != nil {
 		return nil, err
 	}
 	weekday := t.Weekday().String()
 	return ast.String(weekday), nil
 }
 
-func utcTime(a ast.Value) (time.Time, error) {
-	value, err := builtins.NumberOperand(a, 1)
+func tzTime(a ast.Value) (t time.Time, err error) {
+	var nVal ast.Value
+	loc := time.UTC
+
+	switch va := a.(type) {
+	case ast.Array:
+
+		// not sure what error to return if passed an empty array, but if this isn't handled it panics if passed an empty array
```

Comment:
We probably shouldn't cause a golang panic by referencing `va[0]` without checking.  Returning an error seems like a better option, one of these two might be suitable https://github.com/open-policy-agent/opa/blob/master/topdown/builtins/builtins.go#L58-L77 @tsandall might have a better suggestion for which.

Reply:
Thanks, that's what I meant, l was looking for a pointer as to what the error handler should look like if it's going to give an error message that shows all the possible inputs. The variants within `NewOperandElementErr()` don't really handle something with two signatures as far as I can tell.

There should also be a `default:` in the switch to handle being passed something other than a number or an array, and the error would be pretty much the same thing.

The other alternative is to have different functions so they can be declared with strong interfaces (i.e. not `any`) so they don't get passed the wrong types in the first place. So maybe have `time.date()` and `time.date_tz()`?

A more complex alternative would be to extend the builtin handling to be able to handle multiple functions with the same name but different signatures. This would mean that the `builtinFunctions` array would need to be keyed on something other than the name. This actually feels like the cleanest solution to me, being able to explicitly declare strongly typed variations rather than making them be `any` and having them do their own validation, but it's a bit more work, obviously.

Reply:
> A more complex alternative would be to extend the builtin handling to be able to handle multiple functions with the same name but different signatures. 

This can sort of be achieved today with the `types.NewAny()` helper for the arguments and switching in the builtin, similar to what you're doing. You can just have those call separate method implementations.

As far as the `default` case check out the `to_number` declaration  https://github.com/open-policy-agent/opa/blob/master/ast/builtins.go#L528-L544 and implementation https://github.com/open-policy-agent/opa/blob/master/topdown/casts.go#L14

That case and final error returned are probably what we'll want for these. We should switch to using the `types.NewAny()` helper to be more explicit about what types are allowed.

As far as the empty array case I think its ok to just return a new error explaining them, no need to use the helpers if they don't fit the use-case.

Reply:
Added in the error handling, I think it's OK, the error message is clear enough.